hazelcast/src/main/java/com/hazelcast/spi/impl/AbstractInvocationFuture.java

```diff
@@ -0,0 +1,399 @@
+     *
+     * The state field is replaced using a cas, so registration or setting a response is an atomic operation and therefor not
+     * prone to data-races. There is no need to use synchronized blocks.
+     */
+    private volatile Object state = VOID;
+
+    public AbstractInvocationFuture(Executor defaultExecutor, ILogger logger) {
+        this.defaultExecutor = defaultExecutor;
+        this.logger = logger;
+    }
+
+    boolean compareAndSetState(Object oldState, Object newState) {
+        return STATE.compareAndSet(this, oldState, newState);
+    }
+
+    protected final Object getState() {
+        return state;
+    }
+
+    @Override
+    public final boolean isDone() {
+        return isDone(state);
+    }
+
+    private boolean isDone(final Object state) {
+        if (state == null) {
+            return true;
+        }
+
+        return !(state == VOID
+                || state.getClass() == WaitNode.class
+                || state instanceof Thread
+                || state instanceof ExecutionCallback);
+    }
+
+    protected void onInterruptDetected() {
+    }
+
+    @Override
+    public final boolean cancel(boolean mayInterruptIfRunning) {
+        return false;
+    }
+
+    @Override
+    public final boolean isCancelled() {
+        return false;
+    }
+
+    @Override
+    public final V join() {
+        try {
+            // this method is quite inefficient when there is unchecked exception, because it will be wrapped
+            // in a ExecutionException, and then it is unwrapped again.
+            return get();
+        } catch (Throwable throwable) {
+            throw rethrow(throwable);
+        }
+    }
+
+    @Override
+    public final V getSafely() {
+        return join();
+    }
+
+    @Override
+    public final V get() throws InterruptedException, ExecutionException {
+        Object response = registerWaiter(Thread.currentThread(), null);
+        if (response != VOID) {
+            // no registration was done since a value is available.
+            return resolveAndThrow(response);
+        }
+
+        boolean interrupted = false;
+        try {
+            for (; ; ) {
+                park();
+                if (isDone()) {
+                    return resolveAndThrow(state);
+                } else if (Thread.interrupted()) {
+                    interrupted = true;
+                    onInterruptDetected();
+                }
+            }
+        } finally {
+            restoreInterrupt(interrupted);
+        }
+    }
+
+    @Override
+    public final V get(final long timeout, final TimeUnit unit)
+            throws InterruptedException, ExecutionException, TimeoutException {
+        Thread currentThread = Thread.currentThread();
+
+        Object response = registerWaiter(currentThread, null);
+        if (response != VOID) {
+            return resolveAndThrow(response);
+        }
+
+        // todo: what about overflow? Imagine someone calls with Long.MAX_VALUE
```

Comment:
this should be fixed - as I can imagine it's quite possible. 

the same goes for `get(someValue, unit)` where `unit.toNanos(someValue)` causes overflow.